lib/components/position_component.dart

```diff
@@ -0,0 +1,193 @@
+  TextConfig get debugTextConfig => TextConfig(color: debugColor, fontSize: 12);
+
+  Position toPosition() => Position(x, y);
+  void setByPosition(Position position) {
+    x = position.x;
+    y = position.y;
+  }
+
+  Position toSize() => Position(width, height);
+  void setBySize(Position size) {
+    width = size.x;
+    height = size.y;
+  }
+
+  Rect toRect() => Rect.fromLTWH(x - anchor.relativePosition.dx * width,
+      y - anchor.relativePosition.dy * height, width, height);
+  void setByRect(Rect rect) {
+    x = rect.left + anchor.relativePosition.dx * rect.width;
+    y = rect.top + anchor.relativePosition.dy * rect.height;
+    width = rect.width;
+    height = rect.height;
+  }
+
+  double angleBetween(PositionComponent c) {
+    return (atan2(c.x - x, y - c.y) - pi / 2) % (2 * pi);
+  }
+
+  double distance(PositionComponent c) {
+    return sqrt(pow(y - c.y, 2) + pow(x - c.x, 2));
+  }
+
+  void renderDebugMode(Canvas canvas) {
+    canvas.drawRect(Rect.fromLTWH(0.0, 0.0, width, height), _debugPaint);
+    debugTextConfig.render(
+      canvas,
+      "x: ${x.toStringAsFixed(2)} y:${y.toStringAsFixed(2)}",
+      Position(-50, -15),
+    );
+
+    final Rect rect = toRect();
+    final dx = rect.right;
+    final dy = rect.bottom;
+    debugTextConfig.render(
+      canvas,
+      "x:${dx.toStringAsFixed(2)} y:${dy.toStringAsFixed(2)}",
+      Position(width - 50, height),
+    );
+  }
+
+  void prepareCanvas(Canvas canvas) {
+    canvas.translate(x, y);
+
+    canvas.rotate(angle);
+    final double dx = -anchor.relativePosition.dx * width;
+    final double dy = -anchor.relativePosition.dy * height;
+    canvas.translate(dx, dy);
+
+    // Handle inverted rendering by moving center and flipping.
+    if (renderFlipX || renderFlipY) {
+      canvas.translate(width / 2, height / 2);
+      canvas.scale(renderFlipX ? -1.0 : 1.0, renderFlipY ? -1.0 : 1.0);
+      canvas.translate(-width / 2, -height / 2);
+    }
+
+    if (debugMode) {
+      renderDebugMode(canvas);
+    }
+  }
+
+  void addEffect(PositionComponentEffect effect) {
+    _effects.add(effect..component = this);
+  }
+
+  void propagateToChildren<T extends PositionComponent>(
+      void Function(T, Rect) handler) {
+    final rect = toRect();
+    if (this is T) {
+      handler(this as T, rect);
+    }
+    _children.forEach((c) {
+      if (c is PositionComponent) {
+        final newRect = c.toRect().translate(rect.left, rect.top);
+        if (c is T) {
+          handler(this as T, newRect);
+        }
+        c.propagateToChildren(handler);
+      }
+    });
+  }
+
+  @mustCallSuper
+  @override
+  void resize(Size size) {
+    super.resize(size);
+    _children.forEach((child) => child.resize(size));
+  }
+
+  @mustCallSuper
+  @override
+  void render(Canvas canvas) {
+    prepareCanvas(canvas);
+    canvas.save();
+    _children.forEach((comp) => _renderChild(canvas, comp));
+    canvas.restore();
+  }
+
+  void _renderChild(Canvas canvas, Component c) {
```

Comment:
Just wondering if that would handle composed priority?
e.g. you set custom priority for a parent and all it's children are rendered with correct order.

Reply:
they are within each level, not across the tree. so all siblings that are within the same parent when we do _children.forEach it goes by priority because of the ordered_set